### ISSUE TITLE
Unable to remove unused provider

### ISSUE THREAD

**Issue by martin-flaregames** [hashicorp/terraform#18104]
### Terraform Version
```
Terraform v0.11.7
+ provider.aws v1.19.0
+ provider.template v1.0.0
```

### The issue

TL;DR; I cannot remove a provider which is no longer used and has no resources depending on it.

Not sure if this is only affecting the AWS provider (my case), or if this situation is a general one for any provider.

My situation is the following:

I started a project using a single `aws` provider in `us-east-1` without an alias.

```hcl
provider "aws" {
  region = "${var.aws_region}"
  access_key = "${var.aws_access_key}"
  secret_key = "${var.aws_secret_key}"
}
```

But after some months the customer requested that the project should be distributed over differed AWS regions (us/europe/asia).

In order to make the migration I preserved the old provider for managing (and destroying) the old architecture and created 3 new ones for the migration to the new architecture, a planned downtime during migration was an accepted situation, so I decided also to create another the provider in `us-east-1` with a proper alias.

```hcl
# Old default provider (TODO remove after live migration)
provider "aws" {
  region = "${var.aws_region}"
  access_key = "${var.aws_access_key}"
  secret_key = "${var.aws_secret_key}"
}

# US (Virginia)
variable "us_aws_region" { default = "us-east-1"}
variable "us_name" { default = "us"}
provider "aws" {
  alias  = "${var.us_name}"
  region = "${var.us_aws_region}"
  access_key = "${var.aws_access_key}"
  secret_key = "${var.aws_secret_key}"
}

# Europe (Ireland)
variable "europe_aws_region" { default = "eu-west-1"}
variable "europe_name" { default = "europe"}
provider "aws" {
  alias  = "${var.europe_name}"
  region = "${var.europe_aws_region}"
  access_key = "${var.aws_access_key}"
  secret_key = "${var.aws_secret_key}"
}

# Asia (Singapore)
variable "asia_aws_region" { default = "ap-southeast-1"}
variable "asia_name" { default = "asia"}
provider "aws" {
  alias  = "${var.asia_name}"
  region = "${var.asia_aws_region}"
  access_key = "${var.aws_access_key}"
  secret_key = "${var.aws_secret_key}"
}
```
After the migration there are no more resources depending on the legacy provider so I decided to simply remove it from the `.tf` file.

And this is when the problem occurs. Terraform is complaining that the no longer existing provider cannot be configured.

```bash
terraform validate
Error: provider.aws: "region": required field is not set
```

or

```bash
terraform plan
Acquiring state lock. This may take a few moments...
provider.aws.region
  The region where AWS operations will take place. Examples
  are us-east-1, us-west-2, etc.

  Default: us-east-1
  Enter a value: 
```

### Expected Behavior

I should be able to remove the unused provider.

### Actual Behavior

Terraform complains :)

### Steps to Reproduce
```Bash
terraform init
terraform validate
terraform plan
```

**Comment by martin-flaregames** [comment]
Worst case scenario for opening a ticket after hours of looking for a cause and believing to have found a bug: Finding out that there is no issue a few minutes later after submitting it. 

Using `terraform graph` I figured out that a data source (not a resource) was depending on the old provider. I just appended to provider line to it and...

 ```diff
# :: inline policy document
data "aws_iam_policy_document" "iamdoc_ec2role_inline_policy" {
  statement {
    effect = "Allow"
    actions = [
      "ec2:CreateSnapshot",
      "ec2:DescribeInstances",
      "ec2:DescribeVolumes",
      "ec2:CreateTags"
    ]
    resources = [ "*" ]
  }
+  provider = "aws.${var.us_name}"
}
```

... voila! The problem is gone.

I'm still failing to understand why a IAM policy document data source needs a provider, but that's another topic.

Feel free to close this issue.

**Comment by jbardin** [comment]
Hi @martin-flaregames,

Glad you found the source of the issue here.
The `terraform` binary itself has no resources, it essentially just coordinates providers. The data source needs to come from its provider just like any other resource.

**Comment by holms** [comment]
Happy for you to solve it so quick. I've got `provider.postgresql` who doesn't want to go away: 
```* provider.postgresql: Error initializing PostgreSQL client: error detecting capabilities: error PostgreSQL version: dial tcp :5432: getsockopt: connection refused
```
Trying to remove it from state: `terraform state rm provider.postgresql`, it says one record removed, but it still produces same error, tried to remove it again and it always says `removed 1 item`, and that's it.

**Comment by apparentlymart** [comment]
Hi @holms,

`terraform state rm` wants a resource address rather than a provider address. It ought to be producing an error when you give it a provider address instead but I'm betting the parser thinks you are asking for a resource named `postgresql` of type `provider` :roll_eyes: . I believe in the implementation in master that'd be a "no such resource" error, but not in 0.11.x apparently.

The resolution here would be to find which resource in the state is depending on `provider.postgresql` and remove that resource.

You have something pretty different than what this issue was originally about, so if this comment doesn't resolve it please open a new issue and fill out all the information in the issue template and we can try to figure out what's going on.

**Comment by holms** [comment]
@apparentlymart Thank you! Scary thing actually. So creating users in postgres of course fails, because I've failed with my  VPC security groups. And this led to locking me with this error. I wonder how this should be done? Having two environments or workspaces? Or two separate state files? I've had this chicken and egg problems recently. What should I do when I need to setup let's say vault secret manager cluster first, and then to use values from vault. If vault configuration will fail, all other resources will fail because they won't be able to get secrets. Any solution to this problem? The thing is that if we have two state files, than I'd like to re-use maybe certain variables from the project, maybe even certain providers. What's the best case in here?
